File: gfapy/field/integer.py
import gfapy
import re

def decode(string):
  try:
    return int(string)
  except:
    raise gfapy.FormatError("the string does not represent a valid integer")

unsafe_decode = decode

def validate_decoded(obj):
  if isinstance(obj, int):
    pass
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: str, int)")

def validate_encoded(string):
  if not re.match("^[-+]?[0-9]+$", string):
    raise gfapy.FormatError(
      "{} does not represent a valid integer\n".format(repr(string)) +
      "(it does not match the regular expression [-+]?[0-9]+)")

def unsafe_encode(obj):
  return str(obj)

def encode(obj):
  if isinstance(obj, str):
    validate_encoded(obj)
    return obj
  elif isinstance(obj, int):
    return str(obj)
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: str, int)")

File: lib/rgfa/field/integer.rb
module RGFA::Field::Integer

  def decode(string)
    Integer(string) rescue raise RGFA::FormatError,
      "the string does not represent a valid integer"
  end

  alias_method :unsafe_decode, :decode

  def validate_decoded(integer)
    # always valid
  end

  def validate_encoded(string)
    if string !~ /^[-+]?[0-9]+$/
      raise RGFA::FormatError,
        "#{string.inspect} does not represent a valid integer\n"+
        "(it does not match the regular expression [-+]?[0-9]+)"
    end
  end

  def validate(object)
    case object
    when String
      validate_encoded(object)
    when Integer
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: Integer, String)"
    end
  end

  def unsafe_encode(object)
    object.to_s
  end

  def encode(object)
    case object
    when String
      validate_encoded(object)
      return object
    when Integer
      return object.to_s
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: Integer, String)"
    end
  end

  module_function :decode
  module_function :unsafe_decode
  module_function :validate_encoded
  module_function :validate_decoded
  module_function :validate
  module_function :unsafe_encode
  module_function :encode

end
